news-bundle/src/Controller/Page/NewsFeedController.php

```diff
@@ -0,0 +1,104 @@
+use Contao\CoreBundle\DependencyInjection\Attribute\AsPage;
+use Contao\CoreBundle\Routing\Page\DynamicRouteInterface;
+use Contao\CoreBundle\Routing\Page\PageRoute;
+use Contao\NewsBundle\Event\FetchArticlesForFeedEvent;
+use Contao\NewsBundle\Event\TransformArticleForFeedEvent;
+use Contao\PageModel;
+use FeedIo\Feed;
+use FeedIo\Specification;
+use Symfony\Component\HttpFoundation\Request;
+use Symfony\Component\HttpFoundation\Response;
+
+#[AsPage(contentComposition: false)]
+class NewsFeedController extends AbstractController implements DynamicRouteInterface
+{
+    final public const TYPE = 'news_feed';
+
+    public static $contentTypes = [
+        'atom' => 'application/atom+xml',
+        'json' => 'application/feed+json',
+        'rss' => 'application/rss+xml',
+    ];
+
+    private array $urlSuffixes = [
+        'atom' => '.xml',
+        'json' => '.json',
+        'rss' => '.xml',
+    ];
+
+    public function __construct(private readonly ContaoContext $contaoContext, private readonly Specification $specification)
+    {
+    }
+
+    public function __invoke(Request $request, PageModel $pageModel): Response
+    {
+        $this->initializeContaoFramework();
+
+        $staticUrl = $this->contaoContext->getStaticUrl();
+        $baseUrl = $staticUrl ?: $request->getSchemeAndHttpHost();
+
+        $feed = (new Feed())
+            ->setTitle($pageModel->title)
+            ->setDescription($pageModel->description)
+            ->setLanguage($pageModel->language)
+        ;
+
+        $event = new FetchArticlesForFeedEvent($feed, $request, $pageModel);
+
+        $dispatcher = $this->container->get('event_dispatcher');
+        $dispatcher->dispatch($event);
+
+        if (null !== ($articles = $event->getArticles())) {
+            foreach ($articles as $article) {
+                $event = new TransformArticleForFeedEvent($article, $feed, $pageModel, $request, $baseUrl);
+                $dispatcher->dispatch($event);
+
+                $feed->add($event->getItem());
+
+                $this->tagResponse($article);
+                $this->tagResponse('contao.db.tl_news_archive.'.$article->pid);
+            }
+        }
+
+        $formatter = $this->specification->getStandard($pageModel->feedFormat)->getFormatter();
+
+        $response = new Response($formatter->toString($feed));
+        $response->headers->set('Content-Type', self::$contentTypes[$pageModel->feedFormat]);
+
+        $this->setCacheHeaders($response, $pageModel);
+
+        return $response;
+    }
+
+    public function configurePageRoute(PageRoute $route): void
+    {
+        $format = $route->getPageModel()->feedFormat;
+
+        if (!isset($this->urlSuffixes[$format])) {
+            throw new \RuntimeException(sprintf('%s is not a valid format. Must be one of: %s', $format, implode(',', array_keys($this->urlSuffixes))));
+        }
+
+        $route->setUrlSuffix($this->urlSuffixes[$format]);
+    }
+
+    public function getUrlSuffixes(): array
```

Comment:
What's the purpose of this method?

Reply:
See https://github.com/contao/contao/blob/a5c401df53881cd09edd0e5dfc9813dbb659c251/core-bundle/src/Routing/Page/DynamicRouteInterface.php